tests/datamodules/test_experience_sources.py

```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
+        self.env = [env1 for _ in range(1)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent, n_steps=2)
 
-    def test_discounted_transition(self):
-        self.source = NStepExperienceSource(self.env, self.agent, n_steps=3, gamma=0.9)
+        history = self.source.histories[0]
+        history.append(exp1)
 
-        self.source.n_step_buffer.append(self.experience01)
-        self.source.n_step_buffer.append(self.experience02)
-        self.source.n_step_buffer.append(self.experience03)
+        for idx, exp in enumerate(self.source.stepper(self.device)):
 
-        reward, next_state, done = self.source.get_transition_info()
+            self.assertTrue(isinstance(exp, tuple))
 
-        reward_01 = self.experience02.reward + 0.9 * self.experience03.reward * (1 - done)
-        reward_gt = self.experience01.reward + 0.9 * reward_01 * (1 - done)
+            if idx == 0:
+                self.assertTrue(len(exp) == self.source.n_steps)
+            elif idx == 1:
+                self.assertTrue(len(exp) == self.source.n_steps - 1)
+                self.assertTrue(torch.equal(exp[0].new_state, s1))
 
-        self.assertEqual(reward, reward_gt)
-        self.assertEqual(next_state.all(), self.next_state_02.all())
-        self.assertEqual(self.experience03.done, done)
+                break
 
-    def test_multi_step_discount(self):
-        self.source = NStepExperienceSource(self.env, self.agent, n_steps=3, gamma=0.9)
-        self.source.env.step = Mock(return_value=(self.next_state_02, self.reward_02, self.done_02, Mock()))
+    def test_source_is_done_metrics(self):
+        """Test that when done and the history is full, return the full history"""
 
-        self.source.n_step_buffer.append(self.experience01)
-        self.source.n_step_buffer.append(self.experience02)
+        s1 = torch.ones(3)
+        s2 = torch.zeros(3)
+        r = 1.0
+        done = True
+        _ = Mock()
 
-        reward_gt = 1.71
+        exp1 = Experience(state=s1, action=1, reward=r, done=False, new_state=s2)
+        exp2 = Experience(state=s1, action=1, reward=r, done=False, new_state=s2)
 
-        exp, reward, done = self.source.step(self.device)
+        env1 = Mock()
+        env1.step = Mock(return_value=(s1, r, done, _))
 
-        self.assertEqual(exp[0].all(), self.experience01.state.all())
-        self.assertEqual(exp[1], self.experience01.action)
-        self.assertEqual(exp[2], reward_gt)
-        self.assertEqual(exp[3], self.experience02.done)
-        self.assertEqual(exp[4].all(), self.experience02.new_state.all())
+        n_steps = 3
+        n_envs = 2
+
+        self.env = [env1 for _ in range(2)]
+        self.device = torch.device('cpu')
+        self.source = ExperienceSource(self.env, self.agent, n_steps=3)
 
+        history = self.source.histories[0]
+        history.append(exp1)
+        history.append(exp2)
+        history.append(exp2)
```

Comment:
```suggestion
        history += [exp1, exp2, exp2]
```
really twice the same exp2?